src/pydrex/diagnostics.py

```diff
@@ -0,0 +1,193 @@
+import itertools as it
+
+import numpy as np
+from numpy import random as rn
+import scipy.linalg as la
+
+import pydrex.tensors as _tensors
+
+
+def bingham_average(orientations):
+    """Compute bingham averages from olivine orientation matrices.
+
+    Returns a tuple of vectors which are the averaged
+    a-, b- and c-axes orientations, respectively.
+
+    """
+    # https://courses.eas.ualberta.ca/eas421/lecturepages/orientation.html
+    # Eigenvector corresponding to largest eigenvalue is the mean direction.
+    # SciPy returns eigenvalues in ascending order (same order for vectors).
+    a_mean = np.zeros((3, 3))
+    a_mean[0, 0] = np.sum(orientations[:, 0, 0] ** 2)
+    a_mean[1, 1] = np.sum(orientations[:, 0, 1] ** 2)
+    a_mean[2, 2] = np.sum(orientations[:, 0, 2] ** 2)
+    a_mean[0, 1] = np.sum(orientations[:, 0, 0] * orientations[:, 0, 1])
+    a_mean[0, 2] = np.sum(orientations[:, 0, 0] * orientations[:, 0, 2])
+    a_mean[1, 2] = np.sum(orientations[:, 0, 1] * orientations[:, 0, 2])
+    _, a_eigenvectors = la.eigh(_tensors.upper_tri_to_symmetric(a_mean))
+    a_vector = a_eigenvectors[:, -1]
+
+    b_mean = np.zeros((3, 3))
+    b_mean[0, 0] = np.sum(orientations[:, 1, 0] ** 2)
+    b_mean[1, 1] = np.sum(orientations[:, 1, 1] ** 2)
+    b_mean[1, 2] = np.sum(orientations[:, 1, 2] ** 2)
+    b_mean[0, 1] = np.sum(orientations[:, 1, 0] * orientations[:, 1, 1])
+    b_mean[0, 2] = np.sum(orientations[:, 1, 0] * orientations[:, 1, 2])
+    b_mean[1, 2] = np.sum(orientations[:, 1, 1] * orientations[:, 1, 2])
+    _, b_eigenvectors = la.eigh(_tensors.upper_tri_to_symmetric(b_mean))
+    b_vector = b_eigenvectors[:, -1]
+
+    c_mean = np.zeros((3, 3))
+    c_mean[0, 0] = np.sum(orientations[:, 2, 0] ** 2)
+    c_mean[1, 1] = np.sum(orientations[:, 2, 1] ** 2)
+    c_mean[2, 2] = np.sum(orientations[:, 2, 2] ** 2)
+    c_mean[0, 1] = np.sum(orientations[:, 2, 0] * orientations[:, 2, 1])
+    c_mean[0, 2] = np.sum(orientations[:, 2, 0] * orientations[:, 2, 2])
+    c_mean[1, 2] = np.sum(orientations[:, 2, 1] * orientations[:, 2, 2])
+    _, c_eigenvectors = la.eigh(_tensors.upper_tri_to_symmetric(c_mean))
+    c_vector = c_eigenvectors[:, -1]
+
+    return (
+        a_vector / la.norm(a_vector),
+        b_vector / la.norm(b_vector),
+        c_vector / la.norm(c_vector),
+    )
+
+
+def resample_orientations(orientations, fractions, n_samples=None):
```

Comment:
This is an implementation of what Fraters described in the supplementary information section 2:

> The weighted data is the result of a procedure which effectively resamples the aggregate of N grains, by sampling the largest grains a number of times that is proportional to the volume fraction of the grain and leaving the smallest grains unrepresented in the weighted sample. In this procedure the grains are sorted by volume fraction from smallest to largest. The cumulative volume fraction is then calculated
for each grain. Next, a vector of N random numbers with values between 0 and 1 is created. For each random number we determine how many grains have a cumulative volume fraction less than this random number. This number is then used as the index into the sorted list of grains, and the value of the volume fraction and Euler angle for this index is stored. This is repeated for all N random numbers. This procedure reproduces a volumetric random sampling of the aggregate.

If I understand correctly, we basically resample from [the empirical distribution](https://en.wikipedia.org/wiki/Empirical_distribution_function). Therefore upsampling/downsampling just works, because for each sample we just pick a value from the original data based on the weights and return it. I will add tests though for `n_samples != len(fractions)`.